framework/py/flwr/superlink/servicer/control/control_servicer.py

```diff
@@ -96,23 +107,70 @@ def __init__(  # pylint: disable=R0913, R0917
         is_simulation: bool,
         authn_plugin: ControlAuthnPlugin,
         artifact_provider: ArtifactProvider | None = None,
+        fleet_api_type: str | None = None,
     ) -> None:
         self.linkstate_factory = linkstate_factory
         self.ffs_factory = ffs_factory
         self.objectstore_factory = objectstore_factory
         self.is_simulation = is_simulation
         self.authn_plugin = authn_plugin
         self.artifact_provider = artifact_provider
+        self.fleet_api_type = fleet_api_type
 
-    def StartRun(  # pylint: disable=too-many-locals
+    def StartRun(  # pylint: disable=too-many-locals, too-many-branches, too-many-statements
         self, request: StartRunRequest, context: grpc.ServicerContext
     ) -> StartRunResponse:
         """Create run ID."""
         log(INFO, "ControlServicer.StartRun")
         state = self.linkstate_factory.state()
         ffs = self.ffs_factory.ffs()
 
-        if len(request.fab.content) > FAB_MAX_SIZE:
+        verification_dict: dict[str, str] = {}
+        if request.app_spec.startswith("@"):
+            if self.fleet_api_type == TRANSPORT_TYPE_GRPC_ADAPTER:
+                log(
+                    ERROR,
+                    "FAB downloading is not supported when using the gRPC adapter.",
+                )
+                return StartRunResponse()
+
+            # Parse app specification
+            app_id, app_version = _parse_app_spec(request.app_spec)
+
+            # Validate app version format
+            if app_version:
+                if not re.match(APP_VERSION_PATTERN, app_version):
+                    log(
+                        ERROR,
+                        "Invalid app version. Expected format: x.y.z (digits only).",
+                    )
+                    return StartRunResponse()
+
+            # Validate app ID format
+            if not re.match(APP_ID_PATTERN, app_id):
+                log(
+                    ERROR,
+                    "Invalid remote app ID. Expected format: '@account_name/app_name'.",
+                )
+                return StartRunResponse()
+
+            # Request download link
+            url, verifications = _request_download_link(app_id, app_version, context)
+            verification_dict = _format_verification(verifications, verification_dict)
+
+            # Download FAB from Flower platform API
+            try:
+                r = requests.get(url, timeout=60)
+                r.raise_for_status()
+            except requests.RequestException as e:
+                log(ERROR, "FAB download failed: %s", str(e))
+                return StartRunResponse()
+            fab_file = r.content
```

Comment:
should we move this to a separate function? it will be easier to write test for it in that way.